File: backend/src/main/java/com/workwell/run/CompliancePeriod.java
package com.workwell.run;

import java.time.LocalDate;

/**
 * Compliance-cycle bucketing for recurring evaluations (#150 H1).
 *
 * <p>A nightly monitoring run must keep a bounded, current set of cases — one per
 * (employee &times; measure) per compliance cycle, updated idempotently — instead of minting a new
 * {@code evaluation_period} cohort every run (which produced 4,703 perpetually-open cases).
 *
 * <p>Compliance is still computed <em>as of the run date</em>, so the CQL/outcome numbers are
 * unchanged; only the {@code evaluation_period} that keys outcomes + cases is anchored to the
 * measure's current cycle. The anchor is a {@link LocalDate} so the stored {@code evaluation_period}
 * string stays date-shaped and existing parsers keep working.
 */
public final class CompliancePeriod {

    private CompliancePeriod() {
    }

    /** How often a measure's compliance is re-snapshotted into a fresh case cohort. */
    public enum Cadence {
        ANNUAL,
        BIANNUAL,
        SEASONAL
    }

    /**
     * Classify a measure's cadence from its compliance window + seasonal flag. Seasonal (e.g. flu)
     * → season cycle; short windows (&le; 200 days, e.g. the 180-day HbA1c measure) → half-year;
     * everything else (365, the 820-day CMS125, value-based CMS122) → calendar year.
     */
    public static Cadence cadenceFor(int complianceWindowDays, boolean seasonal) {
        if (seasonal) {
            return Cadence.SEASONAL;
        }
        if (complianceWindowDays > 0 && complianceWindowDays <= 200) {
            return Cadence.BIANNUAL;
        }
        return Cadence.ANNUAL;
    }

    /**
     * The stable cycle-anchor date that every run within the same cycle shares, so re-evaluations
     * collapse onto one case per (employee &times; measure &times; cycle).
     */
    public static LocalDate cycleAnchor(Cadence cadence, LocalDate asOf) {
        return switch (cadence) {
            case ANNUAL -> LocalDate.of(asOf.getYear(), 1, 1);
            case BIANNUAL -> asOf.getMonthValue() <= 6
                    ? LocalDate.of(asOf.getYear(), 1, 1)
                    : LocalDate.of(asOf.getYear(), 7, 1);
            // Flu-style season runs Jul–Jun; anchor at Jul 1 of the season's start year.
            case SEASONAL -> asOf.getMonthValue() >= 7
                    ? LocalDate.of(asOf.getYear(), 7, 1)
                    : LocalDate.of(asOf.getYear() - 1, 7, 1);
        };
    }

    /** The cycle-anchor {@code evaluation_period} (date string) for a measure as of a run date. */
    public static String cycleKey(int complianceWindowDays, boolean seasonal, LocalDate asOf) {
        return cycleAnchor(cadenceFor(complianceWindowDays, seasonal), asOf).toString();
    }
}

File: backend-ts/src/run/compliance-period.ts
/**
 * Compliance-cycle bucketing (#150 H1) — TS port of `com.workwell.run.CompliancePeriod`.
 *
 * The case idempotency key is (employee, measure, evaluation_period). When evaluation_period
 * is the raw run date, every nightly run mints a fresh cohort of cases for the same people —
 * the worklist flood (#150 H1). Bucketing the period to the measure's current compliance CYCLE
 * (annual → Jan 1; biannual → Jan 1 / Jul 1; seasonal/flu → the Jul 1 season anchor) makes
 * repeated runs within one cycle idempotent: the same (employee, measure, period) key upserts
 * the existing case instead of inserting a new one.
 *
 * Pure + deterministic — string in, string out, no Date/timezone surprises — and mirrors the
 * Java helper line-for-line so BOTH stacks bucket identically across the #109 cutover.
 */
import { MEASURE_BINDINGS } from "../engine/synthetic/measure-bindings.ts";

export type Cadence = "ANNUAL" | "BIANNUAL" | "SEASONAL";

/** Flu is the only seasonal (Jul–Jun season) measure; everything else is recency-windowed. */
const SEASONAL_MEASURE_IDS = new Set<string>(["flu_vaccine"]);

/** Cadence from the compliance window: a ≤200-day window is biannual, otherwise annual; seasonal overrides both. */
export function cadenceFor(complianceWindowDays: number, seasonal: boolean): Cadence {
  if (seasonal) return "SEASONAL";
  if (complianceWindowDays > 0 && complianceWindowDays <= 200) return "BIANNUAL";
  return "ANNUAL";
}

/**
 * Anchor date (`YYYY-MM-DD`) of the compliance cycle containing `asOf` (also `YYYY-MM-DD`):
 *   ANNUAL   → Jan 1 of asOf's year
 *   BIANNUAL → Jan 1 (Jan–Jun) or Jul 1 (Jul–Dec)
 *   SEASONAL → Jul 1 of the current Jul–Jun season (prior year's Jul 1 when asOf is Jan–Jun)
 */
export function cycleAnchor(cadence: Cadence, asOf: string): string {
  const year = Number(asOf.slice(0, 4));
  const month = Number(asOf.slice(5, 7));
  switch (cadence) {
    case "ANNUAL":
      return `${year}-01-01`;
    case "BIANNUAL":
      return month <= 6 ? `${year}-01-01` : `${year}-07-01`;
    case "SEASONAL":
      return month >= 7 ? `${year}-07-01` : `${year - 1}-07-01`;
  }
}

/** Cycle-anchor key for a window/seasonal flag at `asOf`. */
export function cycleKey(complianceWindowDays: number, seasonal: boolean, asOf: string): string {
  return cycleAnchor(cadenceFor(complianceWindowDays, seasonal), asOf);
}

/**
 * Measure-aware bucket: resolve the measure's compliance window + seasonality from its binding,
 * then return the cycle anchor for `asOf`. Unknown measures fall back to a 365-day annual cycle.
 */
export function bucketPeriodForMeasure(measureId: string, asOf: string): string {
  const window = MEASURE_BINDINGS[measureId]?.complianceWindowDays ?? 365;
  return cycleKey(window, SEASONAL_MEASURE_IDS.has(measureId), asOf);
}
